tests/mr/device/mr_multithreaded_tests.cpp

```diff
@@ -195,34 +196,41 @@ void allocate_loop(rmm::mr::device_memory_resource* mr,
       RMM_CUDA_TRY(cudaEventRecord(event, stream.value()));
       allocations.emplace_back(ptr, size);
     }
+    allocations_ready.notify_one();
   }
+  // Work around for threads going away before cudaEvent has finished async processing
+  cudaEventSynchronize(event);
 }
 
 void deallocate_loop(rmm::mr::device_memory_resource* mr,
                      std::size_t num_allocations,
                      std::list<allocation>& allocations,
                      std::mutex& mtx,
+                     std::condition_variable& allocations_ready,
                      cudaEvent_t& event,
                      rmm::cuda_stream_view stream)
 {
-  for (std::size_t i = 0; i < num_allocations;) {
-    std::lock_guard<std::mutex> lock(mtx);
-    if (allocations.empty()) { continue; }
-    i++;
+  for (std::size_t i = 0; i < num_allocations; i++) {
+    std::unique_lock lock(mtx);
+    allocations_ready.wait(lock, [&allocations] { return !allocations.empty(); });
     RMM_CUDA_TRY(cudaStreamWaitEvent(stream.value(), event));
     allocation alloc = allocations.front();
     allocations.pop_front();
     mr->deallocate(alloc.ptr, alloc.size, stream);
+    lock.unlock();
```

Comment:
Since this is the last statement in the block, it could be removed (since the default destructor is just about to be called which will `unlock` anyway), but meh.